### PR TITLE
[Bug fix] moreh_norm_backward: support p=inf, -inf and 0 (#56248)

### LINKED ISSUE #56248: [Bug] moreh_norm_backward returns all-zero gradients (and garbage values) for p=±inf, where the forward moreh_norm special-cases these orders
# Issue draft — tt-metal #NEW (moreh norm_backward p=±inf)

**Title**: `[Bug] moreh_norm_backward returns all-zero gradients (and garbage values) for p=±inf, where the forward moreh_norm special-cases these orders`

### Summary

`ttnn.operations.moreh.norm_backward` accepts any float `p` — and the matching
forward `moreh_norm` special-cases `p = +inf`, `p = -inf` and `p = 0` in its
factories (`IS_ZERO` / `MINUS_INF` defines, e.g.
`moreh_norm_program_factory_nc_other.cpp:219-225`). The backward factory has
**no such special-casing**: it feeds `p` (and `p - 1`) straight into
`get_floored_p_and_decimal_and_p_is_negative()`
(`moreh_norm_backward_program_factory.cpp:37-44,119-121`), whose
`static_cast<uint32_t>(floored_p)` is undefined behavior for `±inf`
(`[conv.fpint]`), and whose `decimal = p - floor(p)` is `inf - inf = NaN`.

The result is silently wrong gradients:
- `p = +inf`: **every argmax element's gradient is returned as 0.0** (expected:
  `dy` broadcast to the argmax position — PyTorch semantics), and some non-max
  elements come back with garbage values up to `1.5e38` / `inf`.
- `p = -inf`: the argmin gradient is right, but **non-argmin elements keep a
  non-zero gradient** (expected 0), up to the full `dy` magnitude.

### Evidence (ttsim, Blackhole simulator, ttnn 0.78.0 wheel, bf16)

Minimal deterministic reproducer (seed 0), `x = randn(2, 64)`, `dy = randn(2, 1)`,
`p = +inf`, `dim = 1`:

```
row 0: argmax=59
  expected dx[argmax] = -0.4590   (dy = -0.4590)
  ttnn    dx[argmax] = +0.0000
  expected non-zero grads in row: 1;  ttnn non-zero: 0
row 1: argmax=36
  expected dx[argmax] = +0.3359   (dy = -0.3359)
  ttnn    dx[argmax] = +0.0000
  expected non-zero grads in row: 1;  ttnn non-zero: 0
```

Wider sweep (4 shapes × dims 0/1/2, both orders):

| p | shape, dim | max abs err | argmax grads lost | garbage > 1e30 |
|---|---|---|---|---|
| +inf | (4,100) d1 | 1.54e38 | 5 | 1 |
| +inf | (2,64) d1 | 1.13 | 2 | 0 |
| +inf | (4,100) d0 | 1.16e38 | 99 | 3 |
| +inf | (3,5,40) d2 | inf | 16 | 1 |
| -inf | (4,100) d1 | 1.41 | 0 (argmin right) | 0 |
| -inf | (2,64) d1 | 1.13 | 0 | 0 |
| -inf | (4,100) d0 | 3.41 | 0 | 0 |
| -inf | (3,5,40) d2 | 2.47 | 0 | 0 |

Expected gradients from `torch.linalg.vector_norm(x, ord=±inf, dim).backward(dy)`
(subgradient convention: gradient lands on the argmax/argmin element).

### Root cause

`moreh_norm_backward_program_factory.cpp`:
- L37-44 `get_floored_p_and_decimal_and_p_is_negative(float p)`:
  `floor(±inf)=±inf`, `decimal = ±inf ∓ inf = NaN`,
  `static_cast<uint32_t>(±inf)` = UB (on this target wraps to 0 or 2^32-1,
  hence the garbage exponents).
- L119-121: both `p` and `p - 1` go through it; the kernel then computes
  `|x|^(p-1)` and `1/y^p` via `power_tile_with_abs_x_to_cb` /
  `power_and_recip_tile_to_cb` (`moreh_common.hpp:979,1077`) with those
  meaningless exponent args.

For contrast, the forward factories test and special-case exactly these orders:
`moreh_norm_program_factory_{nc,h,w}_other.cpp` set `IS_ZERO` for `p == 0` and
`MINUS_INF` for `p == -inf`, and the kernels
(`moreh_norm_{nc,h,w}_kernel.cpp`) then use `binary_max_tile` /
`negative_tile` instead of the power ladder. `p = +inf` falls out of the default
path as abs+max. The backward side never got the same treatment.

### Notes / suggested fix direction

- The nightly test `test_moreh_norm_backward` (test_moreh_norm.py:574)
  parametrizes `p` only over `[2.0, 2.5, -2.5]` — the ±inf orders are exercised
  only in the *forward* tests, which is presumably why this never surfaced.
- Suggested: mirror the forward's dispatch. Host-side special-case in the
  backward factory for `p == ±inf` (and `p == 0`), emitting the sub-gradient
  directly: `dx = where(x == argmax_along_dim(|x|), dy, 0)` for +inf (and the
  argmin/-negated variant for -inf). On device this can reuse the same
  abs/max/negative tiles the forward already produces.
- Related but distinct issues: #55585 (zero-norm slices → NaN; PR #55598 adds a
  zero guard but does not touch the ±inf decomposition) and #55587 (the same
  floored-p decomposition is UB in `moreh_clip_grad_norm`). This report is the
  norm_backward ±inf mechanism specifically — verified on-device (simulator)
  rather than source-level only.

### Reproducer

```python
import torch, ttnn

torch.manual_seed(0)
device = ttnn.open_device(device_id=0)
D = lambda t: ttnn.from_torch(t.contiguous(), dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)

x = torch.randn(2, 64).bfloat16()
dy = torch.randn(2, 1).bfloat16()

y = torch.linalg.vector_norm(x.float(), ord=float("inf"), dim=1, keepdim=True)
x64 = x.float().requires_grad_(True)
torch.linalg.vector_norm(x64, ord=float("inf"), dim=1, keepdim=True).backward(dy.float())
ref = x64.grad  # gradient on the argmax element only

got = ttnn.to_torch(ttnn.operations.moreh.norm_backward(
    D(x), D(y.bfloat16()), D(dy), p=float("inf"), dim=1)).float()

for r in range(2):
    amax = x.float().abs()[r].argmax().item()
    print(f"row {r} argmax={amax}: expected {ref[r, amax].item():+.4f}, got {got[r, amax].item():+.4f}")
```

Environment: tt-metal ttnn 0.78.0 wheel (commit of 2026-09-11 main is affected,
no inf handling in `moreh_norm_backward_program_factory.cpp`), Blackhole ttsim
simulator, bf16 tensors, TILE layout.

### PR DESCRIPTION
**Pull request by mohmedmm**
Adds dedicated subgradient dispatch for p = +/-inf and p = 0 in moreh_norm_backward. Prevents entering the floored-p power ladder where infinite exponents produce NaN and undefined behavior, matching PyTorch autograd semantics. Fixes #56248.

### CODE REVIEW

**Comment by bbradelTT** [comment]
Closing in favour of https://github.com/tenstorrent/tt-metal/pull/56334
